File: player/src/main/sc/player2021/Starter.java
package sc.player2021;

import jargs.gnu.CmdLineParser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sc.player2021.logic.Logic;
import sc.player.PlayerClient;
import sc.player.IGameHandler;
import sc.shared.SharedConfiguration;

import java.io.File;

/**
 * Hauptklasse des Clients, die über Konsolenargumente gesteuert werden kann.
 * Sie veranlasst eine Verbindung zum Spielserver.
 */
public class Starter {
  private static final Logger logger = LoggerFactory.getLogger(Starter.class);

  public Starter(String host, int port, String reservation, String roomId) {
    // Strategie zuweisen
    IGameHandler logic = new Logic();
    PlayerClient client = new PlayerClient(host, port, logic);

    // einem Spiel beitreten
    if (reservation != null && !reservation.isEmpty()) {
      client.joinPreparedGame(reservation);
    } else if (roomId != null) {
      client.joinGameRoom(roomId);
    } else {
      client.joinAnyGame();
    }
  }

  public static void main(String[] args) {
    System.setProperty("file.encoding", "UTF-8");

    // Parameter definieren
    CmdLineParser parser = new CmdLineParser();
    CmdLineParser.Option hostOption = parser.addStringOption('h', "host");
    CmdLineParser.Option portOption = parser.addIntegerOption('p', "port");
    CmdLineParser.Option reservationOption = parser.addStringOption('r', "reservation");
    CmdLineParser.Option roomOption = parser.addStringOption("room");

    try {
      // Parameter auslesen
      parser.parse(args);
    } catch (CmdLineParser.OptionException e) {
      // bei einem Fehler die Hilfe anzeigen
      showHelp(e.getMessage());
      System.exit(2);
    }

    // Parameter laden
    String host = (String) parser.getOptionValue(hostOption, "localhost");
    int port = (Integer) parser.getOptionValue(portOption, SharedConfiguration.DEFAULT_PORT);
    String reservation = (String) parser.getOptionValue(reservationOption);
    String room = (String) parser.getOptionValue(roomOption);

    // einen neuen client erzeugen
    try {
      new Starter(host, port, reservation, room);
    } catch (Exception e) {
      logger.error("Beim Starten den Clients ist ein Fehler aufgetreten:", e);
      e.printStackTrace();
    }

  }

  private static void showHelp(String errorMsg) {
    String jarName = new File(Starter.class.getProtectionDomain().getCodeSource().getLocation().getFile()).getName();
    System.out.println("\n" + errorMsg);
    System.out.println("\nFolgende Parameter sind erlaubt: \n"
            + "java -jar " + jarName + " [{-h,--host} hostname]\n"
            + "                               [{-p,--port} port]\n"
            + "                               [{-r,--reservation} reservierung]\n"
            + "                               [--room raumnummer]\n");
    System.out.println("Beispiel: \n"
            + "java -jar " + jarName + " --host 127.0.0.1 --port 10500 --reservation 1234\n");
  }

}

File: player/src/sc/player2021/Starter.kt
package sc.player2021

import jargs.gnu.CmdLineParser
import java.io.File
import kotlin.system.exitProcess

/**
 * Dies ist die Starter Datei, die benutzt wird, um Clients mit dem Server zu verbinden.
 * Wichtig ist die `main` Funktion; diese wird bei Start als erstes ausgeführt.
 *
 * @args [args] Eine Liste aller Parameter, mit der das Programm aufgerufen wurde.
 */
fun main(args: Array<String>) {
    val parser = CmdLineParser()

    /**
     * Dies sind die verschiedenen Optionen, die dem Programm von der Kommandozeile aus übergeben werden können.
     * So kann man mit Start des Programms auswählen,
     * auf welchem Server / zu welchem Spiel sich der Client verbinden soll
     */
    val options = mapOf(
            "host" to parser.addStringOption('h', "host"),
            "port" to parser.addIntegerOption('p', "port"),
            "reservation" to parser.addStringOption('r', "reservation")
    )

    /**
     * Es wird geprüft, ob die vorhanden Argumente Sinn machen, also valide Optionen sind.
     * Bei einem Fehler wird eine kleine Hilfe angezeigt.
     */
    try {
        parser.parse(args)
    } catch (e: CmdLineParser.OptionException) {
        showHelp(e.message.orEmpty())
        exitProcess(2)
    }

    /** Die Optionen werden zu sinnvollen Datentypen umgewandelt. */
    val host = parser.getOptionValue(options["host"],"localhost") as String
    val port = parser.getOptionValue(options["port"], "") as Int
    val reservation = parser.getOptionValue(options["reservation"], "") as String

    /** Hier wird der eigentliche Client gestartet, der dann versucht, sich mit dem gegebenen Server zu verbinden. */
    try {
        SimpleClient(host, port, reservation)
    } catch (e: Exception) {
        SimpleClient.logger.error("Beim starten des Clients ist ein Fehler aufgetreten", e)
        e.printStackTrace()
    }
}

/** Eine Hilfsfunktion, die bei falschem Aufruf des Programms anzeigt, wie das Programm zu verwenden ist. */
private fun showHelp(errorMsg: String) {
    val jarName = File(SimpleClient::class.java.protectionDomain.codeSource.location.file).name
    println("\n$errorMsg")
    println("""
        Bitte das Programm mit folgenden Parametern (optional) aufrufen:
        java -jar $jarName [{-h, --host} hostname]
                       [{-p, --port} port]
                       [{-r, --reservation} reservation]

        Beispiel:
        java -jar $jarName --host 127.0.0.1 --port 10500 --reservation 1234
    """.trimIndent())
}

